packages/datetime/src/dateRangeInput.tsx

```diff
@@ -739,10 +749,10 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
                 ? fromMomentToDate(selectedBound)
                 : undefined;
         });
-        // show only the start date if the dates overlap
-        // TODO: add different handling for the === case once
-        // allowSingleDayRange is implemented (#249)
-        return [startDate, (startDate >= endDate) ? null : endDate] as DateRange;
+        const selectedRange = this.props.allowSingleDayRange
+            ? [startDate, (clearTime(startDate) > clearTime(endDate)) ? null : endDate]
+            : [startDate, (clearTime(startDate) >= clearTime(endDate)) ? null : endDate];
```

Comment:
Yeah there's a little duplication between the implementation for the next-selected dates (on click) and the next focused fields on click (and still more duplication/logic overlap for validation logic).

But you're right: the basic way to check this is `!DateUtils.areSameDay(startDate, endDate) && startDate > endDate`. Will edit.